### ISSUE TITLE
linux-asahi-6.4 bluetooth pairing Issues

### ISSUE THREAD

**Issue by jannau** [AsahiLinux/linux#177]
There are multiple reports that linux-asahi-6.4 has regressed pairing of some Bluetooth devices. Reported devices are sofar only  Logitech mices. As far as I'm aware those use BLE for pairing.

I might notice something related: A Logitech MX Anywhere 2 fails to pair using KDE's bt wizard but pairs via bluetoothctl.

This Fedora Asahi [discussion](https://discussion.fedoraproject.org/t/mx-master-3-bluetooth-mouse-doesnt-connect/87072/19) has conflicting reports about devices with the same product name. So either the devices with the same product name behave differently or the issue might be wlan/bt HW module device specific.

If you are affected please report as much information on the bluetooth device (Product name, bt vendor and product ids, if available device firmware version, production date) and attach the output of asahi-diagnose.
For unaffected setups using the same bluetooth device as being reported as affected please report the same.

**Comment by zanfix** [comment]
My device is a Logitech Pebble M350

I found the following information, when connected to another device (not my macbook air)
```
$ cat /sys/class/input/input13/name 
Logitech Pebble
$ cat /sys/class/input/input13/id/bustype 
0005
$ cat /sys/class/input/input13/id/product 
b021
$ cat /sys/class/input/input13/id/vendor 
046d
$ cat /sys/class/input/input13/id/version
0111
```
Please tell me if the information is the correct one or useful

On the macbook air blueetoothctl gives the following error:
```
[bluetooth]# pair E4:49:B0:xx:xx:xx
Attempting to pair with E4:49:B0:xx:xx:xx
[CHG] Device E4:49:B0:xx:xx:xx Connected: yes
Failed to pair: org.bluez.Error.ConnectionAttemptFailed
[CHG] Device E4:49:B0:xx:xx:xx Connected: no
```

When pairing with kde i see the following errors in the journal (journalctl -xe)
```
Aug 23 19:21:10 zanfix-m1 plasmashell[743]: QString::arg: 2 argument(s) missing in org.kde.bluedevilwizard
Aug 23 19:21:10 zanfix-m1 systemd[620]: Started Add Bluetooth Device - Add Bluetooth Device.
░░ Subject: A start job for unit UNIT has finished successfully
░░ Defined-By: systemd
░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
░░ 
░░ A start job for unit UNIT has finished successfully.
░░ 
░░ The job identifier is 383.
Aug 23 19:21:13 zanfix-m1 bluedevil-wizard[2285]: kf.bluezqt: PendingCall Error: "Page Timeout"
```

[asahi-diagnose-20230823-190410.txt](https://github.com/AsahiLinux/linux/files/12421687/asahi-diagnose-20230823-190410.txt)

**Comment by iaguis** [comment]
I've been unable to connect an Xbox Wireless Controller since updating to linux-asahi-6.4. When I try to connect with bluetoothctl I get the following error:

```
Failed to connect: org.bluez.Error.Failed le-connection-abort-by-local
```

This seems to suggest it's using BLE.

Rolling back to asahi-6.3.0-13 fixes the issue.

Device: Xbox Wireless Controller
Vendor: Microsoft Corporation (0C:35:26)

[asahi-diagnose-20230823-191336.txt](https://github.com/AsahiLinux/linux/files/12421720/asahi-diagnose-20230823-191336.txt)

**Comment by nuovodna** [comment]
Unable to connect a Logitech MX Master 3 for Mac to Macbook Air M1 8Gb with latest asahi kernel.

Trying to pair (trust works) with bluetoothclt gives me this error:
`Attempting to pair with D6:B8:A9:xx:xx:xx
[CHG] Device D6:B8:A9:xx:xx:xx Connected: yes
Failed to pair: org.bluez.Error.ConnectionAttemptFailed
[CHG] Device D6:B8:A9:xx:xx:xx Connected: no
`
It fails with kde systemsettings too:
`kf.bluezqt: PendingCall Error: "Page Timeout"
`
[asahi-diagnose-20230824-001157.txt](https://github.com/AsahiLinux/linux/files/12423678/asahi-diagnose-20230824-001157.txt)

**Comment by PokeCaptain** [comment]
I encountered this bug last night with my keyboard and mouse. Just finished repro-ing for the logs and my mouse surprisingly paired successfully while my keyboard still didn't. Details below:

**Logitech K850**
- BT chip: Nordic Semiconductor N51802 QFAAA0 https://www.nordicsemi.com/Products/nRF51802
- Product ID: M/N: Y-R0061, P/N: 820-008064
- Device FW Version: MPK04.03_0015

**Logitech M720**
- BT chip: Nordic Semiconductor N51802 QFAAA1 https://www.nordicsemi.com/Products/nRF51802
- Product ID: M/N: M-R0062, P/N: 810-006500
- Device FW Version: MPM04.10_0013

[asahi-diagnose-20230825-162302.txt](https://github.com/AsahiLinux/linux/files/12443341/asahi-diagnose-20230825-162302.txt)

**Comment by jasoneckert** [comment]
Up until today I've had zero Bluetooth issues with my Logitech MX Master 3 mouse - I originally paired it using KDE Settings after installing Fedora Asahi Remix on August 10, and it worked fine through all subsequent software installation and updates (including the latest one I did yesterday).

Today, however, I logged into KDE Plasma (I normally use sway) and both my Logitech MX Master 3 and my Lenovo ThinkPad TrackPoint II keyboard (also paired with Bluetooth) stopped working.

After connecting a USB keyboard/mouse and attempting to pair them again using KDE Settings, the process for both failed, adding the following entry to the Systemd journal (twice):

> Aug 27 15:41:32 fedora-asahi bluedevil-wizard[9364]: kf.bluezqt: PendingCall Error: "Page Timeout"

They also failed to pair using `bluetoothctl` as shown here for the MX Master 3:
> [bluetooth]# trust CF:02:2A:xx:xx:xx
> Changing CF:02:2A:xx:xx:xx trust succeeded
> [bluetooth]# pair CF:02:2A:xx:xx:xx
> Attempting to pair with CF:02:2A:xx:xx:xx
> [CHG] Device CF:02:2A:xx:xx:xx Connected: yes
> Failed to pair: org.bluez.Error.ConnectionAttemptFailed
> [CHG] Device CF:02:2A:xx:xx:xx Connected: no
> [bluetooth]# _

Both work fine in macOS, so I tried extracting the appropriate keys from macOS for the MX Master 3 as outlined [here](https://wiki.archlinux.org/title/Bluetooth#Problems_with_all_BLE_devices_on_kernel_5.9+), added them to the IdentityResolvingKey.Key, SlaveLongTermKey.Key, and PeripheralLongTermKey.Key files in the /var/lib/bluetooth/*BTcontroller_MAC_address*/*MXMaster3_MAC_address*/ directory, edited the associated info file in that directory accordingly, restarted the Bluetooth daemon, but that didn't work either.

I've attached the output of `asahi-diagnose` ([asahi-diagnose-20230827-154624.txt](https://github.com/AsahiLinux/linux/files/12448830/asahi-diagnose-20230827-154624.txt)). Additionally, following is the device information regarding the MX Master 3 and TrackPoint II keyboard from macOS (in case the two are related somehow given the symptoms):
>  MX Master 3:
>  Address:	CF:02:2A:xx:xx:xx
>  Vendor ID:	0x046D
>  Product ID:	0xB023
>  Battery Level:	100%
>  Firmware Version:	MPM19.000_0013
>  Minor Type:	Mouse
>  Services:	0x400000 < BLE >
>
>  TrackPoint Keyboard II:
>  Address:	DE:B8:8E:xx:xx:xx
>  Vendor ID:	0x17EF
>  Product ID:	0x60E1
>  Minor Type:	Keyboard
>  Services:	0x400000 < BLE >

***Edit:*** The ThinkPad TrackPoint II keyboard works fine with the Lenovo receiver nub, but interestingly enough, the MX Master 3 doesn’t even work with the Logitech Unifying Receiver. It’s detected, and shows up in KDE Settings as an input device, but doesn’t work at all - yet it works fine with the same Unifying Receiver on a separate Windows system and on macOS too.

***Edit:*** After updating to bluez-5.69 today, the behaviour of `bluetoothctl` changed - the device enters a rapid connect/disconnect loop until you manually `exit`:
> [bluetooth]# trust CF:02:2A:xx:xx:xx
> Changing CF:02:2A:xx:xx:xx trust succeeded
> [bluetooth]# pair CF:02:2A:xx:xx:xx
> Attempting to pair with CF:02:2A:xx:xx:xx
> [CHG] Device CF:02:2A:xx:xx:xx Connected: yes
> Failed to pair: org.bluez.Error.Failed
> [CHG] Device CF:02:2A:xx:xx:xx Connected: no
> [CHG] Device CF:02:2A:xx:xx:xx Connected: yes
> [CHG] Device CF:02:2A:xx:xx:xx Connected: no
> [CHG] Device CF:02:2A:xx:xx:xx Connected: yes
> [CHG] Device CF:02:2A:xx:xx:xx Connected: no
> [CHG] Device CF:02:2A:xx:xx:xx Connected: yes
> [CHG] Device CF:02:2A:xx:xx:xx Connected: no
> [CHG] Device CF:02:2A:xx:xx:xx Connected: yes
> [CHG] Device CF:02:2A:xx:xx:xx Connected: no
> [CHG] Device CF:02:2A:xx:xx:xx Connected: yes
> [bluetooth]# exit

Similarly, when pairing using KDE Settings, it [rapidly appears and disappears](https://triosdevelopers.com/jason.eckert/stuff/MXMaster.mp4) continuously in the list of successfully-paired devices.

**Comment by vklachkov** [comment]
Hello! Some people has the same problems in other distros. Here is an issue from NixOS: https://github.com/NixOS/nixpkgs/issues/242313. It looks like a Linux bug, because on the 6.3 kernel everything works perfectly. It is worth updating to 6.4+, the devices do not pair or do not work...

**Comment by iaguis** [comment]
Reading the nixos issue and the linked bugzilla bug this commit might be the fix: https://github.com/torvalds/linux/commit/95b7015433053

**Comment by iaguis** [comment]
I cherry-picked that commit on top of asahi-6.4-10 and it doesn't fix the issue for me with the Xbox Controller :(

**Comment by jannau** [comment]
> Reading the nixos issue and the linked bugzilla bug this commit might be the fix: [torvalds@95b7015433053](https://github.com/torvalds/linux/commit/95b7015433053)

No, this won't help. It is for the btusb driver this is unrelated the pci hci_bcm4377 driver used on apple silicon devices.

**Comment by jannau** [comment]
On BRCM4378 (M1 devices and at least the M2 Macbook Pro 13-inch (2022) a Logitech MX Anywhere 2 connects and works after 2-3 connection tries. No such luck on M1 Pro/Max/Ultra devices with BRCM4387. No obvious difference in btmon except that one case works and the other one fails. If there is anything suspicious in the btmon log is that there are BLE advertisements during the failed pairing.

**Comment by jannau** [comment]
Reverting 288c90224eec allows pairing of my Logitech MX Anywhere 2 on a Macbook Pro 14-inch (2021). I'm currently looking if there is anything to fix this issue in hci_bcm4377.c

Apparently the same commit regressed certain Intel bt adapters as well and there's now a quirk we can use as well: https://lore.kernel.org/all/20230823194128.3893244-1-luiz.dentz@gmail.com/

**Comment by jannau** [comment]
Will be fixed by #194

**Comment by jannau** [comment]
this is now fixed with kernels based on asahi-6.5-8 and later
